# homework
go训练营作业

## 第二周 error 

`问题： 我们在数据库操作的时候，比如 dao 层中当遇到一个 sql.ErrNoRows 的时候，是否应该 Wrap 这个 error，抛给上层。为什么，应该怎么做请写出代码？`

```go
// ErrNoRows is returned by Scan when QueryRow doesn't return a
// row. In such a case, QueryRow returns a placeholder *Row value that
// defers this error until a Scan.
var ErrNoRows = errors.New("sql: no rows in result set")
```

- 很多人的答案都是需要返回上层，但我最开始看视频听到这个问题的时候，反应很简单，ErrNoRows 不就是没查到吗？不就和 k, ok := m["xxx"] （map）没找到结果一样吗？为什么要返回，这算不上是什么致命的错误啊

- 分场景

  - 如果这个函数的语义是给一个输入，必须输出一个结果（对应的过滤发生在上层），那所有意外的情况都需要返回，因为你只做查一条数据这么简单的一件事，查得到返回正常值，查不到需要返回 err，原因可能是上层过滤出了问题，出现不该有的输入，或者数据库出了问题（但我认为是不会出现的，整个程序已经上线跑起来了，不应该突然的抽风，要挂整个程序挂嘛，怎么会突然挂你一个，而且数据库发生异常了，返回的也不是 ErrNoRows 吧）
  
  - 如果所有的过滤和查询都封装在一个函数中，对外只返回一个结果和一个 bool，那应当就认为所有的错误在内部处理，而 ErrNoRows 我认为不应该算是一个错误，仅仅是查不到值嘛，最好是拿个 bool 来表示，但因为还会有其他类型的错误发生，就把他归为了一个错误来返回，所以说 ErrNoRows 是很正常的一个东西，在内部处理打点日志，如果出现多次，那可能是 bug 或者 恶意请求


- 总结：我目前认为 ErrNoRows 是一个很正常的输出，而不是把他看成一个错误 err，而这个 ErrNoRows 在不同粒度划分的场景下来判断内部处理，还是返回上层。如果在我们说的微服务中，而且划分的非常细，我这个函数只管查一条数据，而且语义上必须是能查得到的，那我查不到了当然要返回上层；如果划分没那么细，我这时候返回一个空值也没问题，那不就内部处理了，所有功能正常的情况下，做点日志，来看是 恶意请求 还是 bug，如果说 ErrNoRows 是由数据库（外部非代码因素）出问题导致的，那应该有其他的程序去监控，仅由 ErrNoRows 怎么判断的出来，只能打打日志这样子

- 再总结：ErrNoRows 这个东西我目前认为不是什么致命且很能反映问题位置的一个 “error”，在强语义要求的函数中，返回错误，在其他地方，内部处理，打日志

```go
// 语义上 进来的 userId 肯定是正确的记录，那么查不到的时候就必须上报是什么原因导致的 
func queryUserInfo(userId int) (userInfo, error) {
	// sql.QueryRow()
	// ...
	
	// 找得到，返回 结果, nil
	
	// 找不到，那可能是上层过滤除了问题，进来了不该有的数据
	// 或者 内部代码 / 数据库 出现问题
	// 而 代码问题 在开发过程中还可能出现
	// 数据库 问题，也不应该在这里反映，有专门的监控
	
	// 总之，必须上报，语义规定的
}
```

```go
// 这条记录可能有也可能没有，而没有的时候，在语义上，我们不需要对外负责，只返回 false
func querySession(session string) (userId, bool) {
	// sql.QueryRow()
	// ...

	// 找得到，返回 结果, true

	// 找不到，可能是 参数问题，pass

	// 或者 内部代码 / 数据库 出现问题
	// 而 代码问题 在开发过程中还可能出现，pass
	// 数据库 问题，也不应该在这里反映，有专门的监控，pass

	// 所以，这个场景出现找不到情况的时候，对于外部，查不到我们不需要负责，在内部，ErrNoRows 这种级别仅仅打点日志足够
}
```

